handel/tests/mod.rs

```diff
@@ -210,227 +208,290 @@ struct NetworkWrapper<N: NetworkInterface<PeerId = MockPeerId>>(Arc<N>);
-            ),
-            NetworkWrapper(net),
+            num_contributors,
+            node_id,
+            &mut hub,
+            tx.clone(),
         );
 
-        let r = stopped.clone();
-        let s = sender.clone();
-        spawn(async move {
-            // have them just run until the aggregation is finished
-            while let Some(contribution) = aggregation.next().await {
-                if contribution.num_contributors() == contributor_num + 1 {
-                    s.send(()).await.expect("Send should never fail");
-                }
-
-                if *r.read() {
-                    return;
-                }
-            }
-        });
+        // Connect the network to all already existing ones.
+        for net in &networks {
+            network.dial_mock(net);
+        }
+        networks.push(network);
     }
 
-    // same as in the for loop, except we want to keep the handel instance and not spawn it.
-    let net = Arc::new(hub.new_network_with_address(contributor_num as u64));
-    let protocol = Protocol::new(contributor_num, contributor_num + 1, contributor_num + 1);
-    let mut contributors = BitSet::new();
-    contributors.insert(contributor_num);
-    let contribution = Contribution {
-        value: contributor_num as u64 + 1u64,
-        contributors,
-    };
-    for network in &networks {
-        net.dial_mock(network);
-    }
+    // Wait for all aggregations to collect `num_contributors - 1` contributions.
+    wait_for_contributions(num_contributors - 1, num_contributors - 1, &mut rx).await;
 
-    // instead of spawning the aggregation task await its result here.
-    let mut aggregation = Aggregation::new(
-        protocol,
+    // Spawn the late node.
+    let network = spawn_handel_instance(
         config.clone(),
-        contribution,
-        Box::pin(
-            net.receive_messages::<Update<Contribution>>()
-                .map(move |msg| msg.0 .0.into_level_update(msg.1 .0 as u16)),
-        ),
-        NetworkWrapper(Arc::clone(&net)),
+        num_contributors,
+        num_contributors - 1,
+        &mut hub,
+        tx.clone(),
     );
 
-    // aggregating should not take more than 1s per each 7 contributors
-    let timeout_ms = 1000u64 * (contributor_num / 7 + 1) as u64;
+    // Connect the network to all already existing ones.
+    for net in &networks {
+        network.dial_mock(net);
+    }
 
-    let deadline = Instant::now()
-        .checked_add(Duration::from_millis(timeout_ms))
-        .unwrap();
+    // Wait for all aggregations to complete.
+    wait_for_contributions(num_contributors, num_contributors, &mut rx).await;
+}
 
-    // The final value needs to be the sum of all contributions.
-    // For instance for `contributor_num = 7: 8 + 7 + 6 + 5 + 4 + 3 + 2 + 1 = 36`
-    let mut exp_agg_value = 0u64;
-    for i in 0..=contributor_num {
-        exp_agg_value += i as u64 + 1u64;
-    }
+#[test(tokio::test)]
+#[ignore]
```

Comment:
Why is this test `#[ignore]`d?

Reply:
So currently the requirement is that validators are somewhat well connected amongst themselves. The `LevelUpdates` are send in a direct fashion. This test connects validators as a straight line, with every validator only be connected to its predecessor. Basically it is a very poorly connected graph. 

As it stands Handel cannot deal with this. The information required cannot be send in such a graph. There is a design for a future upgrade to Handel which will enable such scenarios to resolve. Once we get to implementing that design we will use this test to verify its efficacy. Until then the mitigation is to wait until the network connects better amongst itself, which our network should constantly try to do anyways.